lib/version.c

```diff
@@ -496,11 +502,69 @@ static curl_version_info_data version_info = {
   0,    /* zstd_ver_num */
   NULL, /* zstd version */
   NULL, /* Hyper version */
-  NULL  /* gsasl version */
+  NULL, /* gsasl version */
+  NULL  /* feature_names */
+};
+
+/* Public CURL_VERSION_* feature symbol to name conversion. */
+
+struct feat {
+  const char *name;
+  int bitmask;
+};
+
+static const struct feat feats[] = {
+  {"AsynchDNS",      CURL_VERSION_ASYNCHDNS},
+  {"Debug",          CURL_VERSION_DEBUG},
+  {"TrackMemory",    CURL_VERSION_CURLDEBUG},
+  {"IDN",            CURL_VERSION_IDN},
+  {"IPv6",           CURL_VERSION_IPV6},
+  {"Largefile",      CURL_VERSION_LARGEFILE},
+  {"Unicode",        CURL_VERSION_UNICODE},
+  {"SSPI",           CURL_VERSION_SSPI},
+  {"GSS-API",        CURL_VERSION_GSSAPI},
+  {"Kerberos",       CURL_VERSION_KERBEROS5},
+  {"SPNEGO",         CURL_VERSION_SPNEGO},
+  {"NTLM",           CURL_VERSION_NTLM},
+  {"NTLM_WB",        CURL_VERSION_NTLM_WB},
+  {"SSL",            CURL_VERSION_SSL},
+  {"libz",           CURL_VERSION_LIBZ},
+  {"brotli",         CURL_VERSION_BROTLI},
+  {"zstd",           CURL_VERSION_ZSTD},
+  {"CharConv",       CURL_VERSION_CONV},
+  {"TLS-SRP",        CURL_VERSION_TLSAUTH_SRP},
+  {"HTTP2",          CURL_VERSION_HTTP2},
+  {"HTTP3",          CURL_VERSION_HTTP3},
+  {"UnixSockets",    CURL_VERSION_UNIX_SOCKETS},
+  {"HTTPS-proxy",    CURL_VERSION_HTTPS_PROXY},
+  {"MultiSSL",       CURL_VERSION_MULTI_SSL},
+  {"PSL",            CURL_VERSION_PSL},
+  {"alt-svc",        CURL_VERSION_ALTSVC},
+  {"HSTS",           CURL_VERSION_HSTS},
+  {"gsasl",          CURL_VERSION_GSASL},
+  {"threadsafe",     CURL_VERSION_THREADSAFE},
+  {NULL,             0}
 };
 
+static int strcasecmp4sort(const void *p1, const void *p2)
+{
+#if defined(HAVE_STRCASECMP)
+  return strcasecmp(*(char * const *) p1, *(char * const *) p2);
+#elif defined(HAVE_STRCMPI)
+  return strcmpi(*(char * const *) p1, *(char * const *) p2);
+#elif defined(HAVE_STRICMP)
+  return stricmp(*(char * const *) p1, *(char * const *) p2);
+#else
+  return strcmp(*(char * const *) p1, *(char * const *) p2);
+#endif
```

Comment:
Not a fan of using these functions in libcurl. We use our own functions all over for a reason: they are stable and independent of the locales. This set is not.

I would rather that we provide a version of our internal one with a return code that is fine to sue for qsort.

Reply:
But otoh: why not just have the list in a sorted order to begin with to avoid the qsort?

Reply:
>  I would rather that we provide a version of our internal one with a return code that is fine to sue for qsort.

I was hesitating. I'll create a function in strcase.c for that.
> But otoh: why not just have the list in a sorted order to begin with to avoid the qsort?

Because next features wont have a `CURL_VERSION_*` bitmask and thus would be entered in array by another loop.
I plan to make features for `MIME` and `form-API` in a very near future. Maybe `headers-API` could be one too.

Reply:
> Because next features wont have a CURL_VERSION_* bitmask and thus would be entered in array by another loop.

Ok, I don't think I miss something because I don't see the point in doing additional ones in a second loop.

>  I plan to make features for MIME and form-API in a very near future. Maybe headers-API could be one too.

Why add those? I can't recall anyone ever having asked for for missed them being present there.

Reply:
The #ifdefs could also be used *in* the list, so that disabled features also removes the strings from memory.

Reply:
> Ok, I think I miss something because I don't see the point in doing additional ones in a second loop.

Because they don't have a `CURL_VERSION_*` bitmask.

> Why add those? I can't recall anyone ever having asked for for missed them being present there.

I ran "make check" on a build with `--disable-mime` and several tests fail.
I plan also to really deprecate the form api using a `CURL_DISABLE_FORM_API` conditional and tests for it should check the feature.

> The #ifdefs could also be used in the list, so that disabled features also remove the strings from memory.

The loop introduced in this PR checks the `CURL_VERSION_*` bits in `features` field because some of them are added at run-time. That cannot be done for features without a bitmask.

Alternative: if the bitmask in the table is 0, always include the feature in the array. That could be feasible.

Anyway suppressing the sort would disallow determining the presence of a new feature at run-time and still have the array in alpha order.

Reply:
> The loop introduced in this PR checks the CURL_VERSION_* bits in features field because some of them are added at run-time. That cannot be done for features without a bitmask.

Features that are disabled at build-time don't need their strings as features so they could be removed with #ifdef use. In the spirit of removing things from occupying space when the are disabled.

Reply:
> Features that are disabled at build-time don't need their strings as features so they could be removed with #ifdef use. 

Sure! But that will duplicate all conditional stuff of `version_info.features` initialisation for current features. The idea behind having 2 loops and 2 tables was to avoid this duplication by testing the bitmask for current features in the first loop and to have a compile-time conditioned table for the new ones. I think they may be combined, but I would retain unconditioned entries with a bitmask to avoid conditions duplication.

Reply:
I just pushed a version that:

- has the `feature_names_table` sorted
- has all strings for existing features
- supports (conditioned) new features in single table.
- has no qsort.
- introduces `MIME` feature (as an example).

Reply:
We have several users who are extremely memory and foot print conscious. They .build their own libcurl with only a small set of features and protocols enabled.

For those users, it is important that every feature "carries its own weight" so that when the feature is disabled, as much as possible of that feature is gone and not present in the build. For example: it does not make sense to include the string "http3" in a build that does not have the feature enabled. When we can easily avoid it.

Reply:
> For those users, it is important that every feature "carries its own weight" so that when the feature is disabled, as much as possible of that feature is gone and not present in the build.

I'm aware of that and, as you might have understood, avoiding duplication of conditionals is only motivated by not introducing and error-prone context.
I have an idea for combining both wishes at the expense of a extra function pointer in each compiled-in table entry... let me the time to code it!

Reply:
The last version pushed should satisfy everything at this point.